src/text-editor.coffee

```diff
@@ -2236,6 +2236,60 @@ class TextEditor extends Model
     return unless @getSoftTabs()
     @scanInBufferRange /\t/g, bufferRange, ({replace}) => replace(@getTabText())
 
+  # Essential: Set the indentation level for the given buffer row.
+  #
+  # Rebuilds the indentation based on the soft tabs and tab length settings of
+  # this editor in order to bring it to the given indentation level.
+  # Note that if soft tabs are enabled and the tab length is 2, a row with 4
+  # leading spaces would have an indentation level of 2.
+  #
+  # * `bufferRow` A {Number} indicating the buffer row.
+  # * `newLevel` A {Number} indicating the new indentation level.
+  setIndentLevelForRow: (bufferRow, newLevel) ->
+    @modifyIndentLevelForRow(bufferRow, newLevel)
+
+  # Essential: Adjust the indentation level for the given buffer row.
+  #
+  # Rebuilds the indentation based on the soft tabs and tab length settings of
+  # this editor in order to increase or decrease it by the requested number of
+  # levels.
+  # Note that if soft tabs are enabled and the tab length is 2, a row with 4
+  # leading spaces would have an indentation level of 2.
+  #
+  # * `bufferRow` A {Number} indicating the buffer row.
+  # * `adjustLevel` A {Number} indicating how many levels to increase by. A
+  #   negative value will remove levels.
+  adjustIndentLevelForRow: (bufferRow, adjustLevel) ->
+    @modifyIndentLevelForRow(bufferRow, undefined, adjustLevel)
+
+  # Helper for modifying indentation for a buffer row
+  modifyIndentLevelForRow: (bufferRow, newLevel=undefined, indentIncrease=0) ->
+    leadingWhitespace = @lineTextForBufferRow(bufferRow).match(/^\s*/)[0]
+    newIndentString = @modifyIndentWhitespace(leadingWhitespace, newLevel, indentIncrease)
+    @buffer.setTextInRange([[bufferRow, 0], [bufferRow, leadingWhitespace.length]], newIndentString)
+
+  # Helper for modifying indentation for a line of text
+  modifyIndentLevelForLine: (line, newLevel=undefined, indentIncrease=0) ->
+    line.replace /^\s*/, (matchText) =>
+      @modifyIndentWhitespace(matchText, newLevel, indentIncrease)
+
+  # Helper for modifying indentation whitespace, while preserving straggling
+  # spaces (such as those used to align asterisks in /* */ blocks)
+  modifyIndentWhitespace: (leadingWhitespace, newLevel=undefined, indentIncrease=0) ->
+    {indentLevel, excessSpace} = @displayBuffer.parseIndentation(leadingWhitespace)
+
+    if newLevel?
+      newIndentString = @buildIndentString(newLevel + indentIncrease)
+    else
+      newIndentString = @buildIndentString(indentLevel + indentIncrease)
+
+    # Handle case where spaces were used for alignment (such as /* */ blocks)
+    # immediately after the indentation
+    if excessSpace == 0
```

Comment:
Recommendation: use `is` instead of `==`. See [this section](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#coffeescript-styleguide) of the contributing guide.